jitstats.lua

```diff
@@ -0,0 +1,462 @@
+    statstbl.side_aborts = 0  
+    statstbl.side_completed = 0
+    statstbl.nyi_bc = {}
+    statstbl.exits = 0
+    statstbl.traceflush = 0
+    statstbl.func_blacklisted = 0
+    statstbl.loop_blacklisted = 0
+end
+
+local function table_copy(src, dst)
+    dst = dst or {}
+    for k,v in pairs(src) do
+        if type(v) ~= "table" then
+            dst[k] = v
+        else
+            dst[k] = table_copy(v, dst[k])
+        end
+    end
+    return dst
+end
+
+local function initorclear(tab, key)
+    tab[key] = tab[key] or {}
+    table.clear(tab[key])
+end
+
+--Optionally pass in a table to copy to snapshot to instead of creating a new table
+function jitstats.getsnapshot(dst)
+    if dst then
+        initorclear(dst, "nyi_bc")
+        initorclear(dst, "abort_counts")
+        initorclear(dst, "linktypes")
+    end
+
+    return table_copy(stats, dst)
+end
+
+local function diffstatbl(old, new, result, key)
+    assert(old[key], "missing old sub table")
+    assert(new[key], "missing new sub table")
+    assert(result[key], "missing old sub table")
+    old = old[key]
+    new = new[key]
+    result = result[key]
+
+    for k, v in pairs(new) do
+        local change = v - (old[k] or 0)
+        if change ~= 0 then
+            result[k] = change
+        end
+    end
+    return result
+end
+
+-- Get the diff between two jit stat snapshots
+function jitstats.diffsnapshots(snap1, snap2)
+    local result = {
+        abort_counts = {},
+        nyi_bc = {},
+        linktypes = {},
+    }
+
+    for k, v in pairs(snap2) do   
+        if type(v) == "number" then
+            result[k] = v - (snap1[k] or 0)
+        end
+    end
+
+    diffstatbl(snap1, snap2, result, "abort_counts")
+    diffstatbl(snap1, snap2, result, "nyi_bc")
+    diffstatbl(snap1, snap2, result, "linktypes")
+
+    return result
+end
+
+local last_starts, last_exits
+
+function jitstats.stats_changed()
+
+    if last_starts == stats.starts and last_exits == stats.exits then
+        return false
+    end
+
+    last_starts = stats.starts
+    last_exits = stats.exits
+    return true
+end
+
+function jitstats.setdebugprint(enable)
+    jitstats.debug = enable
+    debugprint = (enable and printf) or function() end
+end
+
+local function topercent(total, subval)
+    return math.floor(subval / total * 100)
+end
+
+function jitstats.print_simple(msg)
+    local running = jitstats.stop()
+    printf("%sStarted %d, Aborted %d, Exits %d", msg or "", stats.starts, stats.aborts, stats.exits)
+    if running then
+        jitstats.start()
+    end
+end
+
+local function table_keys(t)
+    local result = {}
+    for k,_ in pairs(t) do
+        table.insert(result, k)
+    end
+    return result
+end
+
+local function default_sorter(t, k1, k2)
+    return t[k1] > t[k2] 
+end
+
+local function table_sortkeys(t, sorter)
+    local keys = table_keys(t) 
+    sorter = sorter or default_sorter
+    table.sort(keys, function(k1, k2) return sorter(t, k1, k2) end)
```

Comment:
YAGNI?  just do `table.sort(keys, function(k1, k2) return t[k1]>t[k2] end)` and remove the `sorter` parameter

Reply:
I was going to use it to sort by key as well but idk if there would be much use in that.

Reply:
fixed i just simplified it for its single use